fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/rule/TextMatchBasedRewriteRule.java

```diff
@@ -0,0 +1,366 @@
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static com.starrocks.sql.optimizer.OptimizerTraceUtil.logMVRewrite;
+import static com.starrocks.sql.optimizer.rule.transformation.materialization.MaterializedViewRewriter.REWRITE_SUCCESS;
+
+/**
+ * A rule that performs text match based rewrite on a query using materialized views.
+ */
+public class TextMatchBasedRewriteRule extends Rule {
+    private static final Logger LOG = LogManager.getLogger(TextMatchBasedRewriteRule.class);
+
+    private final ConnectContext connectContext;
+    private final StatementBase stmt;
+    private final Map<Operator, ParseNode> optToAstMap;
+
+    // To avoid text match costing too much time, use parameters below to limit it.
+    // limit for sub-query text match(default 4), no match when it <= 0
+    private final int mvSubQueryTextMatchMaxCount;
+    // limit for mvs which matched input query(default 64)
+    private final long mvRewriteRelatedMVsLimit;
+
+    private int subQueryTextMatchCount = 1;
+
+    public TextMatchBasedRewriteRule(ConnectContext connectContext,
+                                     StatementBase stmt,
+                                     Map<Operator, ParseNode> optToAstMap) {
+        super(RuleType.TF_MV_TEXT_MATCH_REWRITE_RULE, Pattern.create(OperatorType.PATTERN));
+
+        this.connectContext = connectContext;
+        this.stmt = stmt;
+        this.optToAstMap = optToAstMap;
+        this.mvSubQueryTextMatchMaxCount =
+                connectContext.getSessionVariable().getMaterializedViewSubQueryTextMatchMaxCount();
+        this.mvRewriteRelatedMVsLimit =
+                connectContext.getSessionVariable().getCboMaterializedViewRewriteRelatedMVsLimit();
+    }
+
+    @Override
+    public List<OptExpression> transform(OptExpression input, OptimizerContext context) {
+        OptExpression rewritten = doTransform(context, input, stmt);
+        return (rewritten != null) ? ImmutableList.of(rewritten) : ImmutableList.of(input);
+    }
+
+    private OptExpression doTransform(OptimizerContext context,
+                                      OptExpression input,
+                                      ParseNode parseNode) {
+        if (context.getOptimizerConfig().isRuleDisable(RuleType.TF_MV_TEXT_MATCH_REWRITE_RULE)) {
+            return null;
+        }
+        SessionVariable sessionVariable = connectContext.getSessionVariable();
+        if (!sessionVariable.isEnableMaterializedViewRewrite() ||
+                !sessionVariable.isEnableMaterializedViewTextMatchRewrite()) {
+            return null;
+        }
+        if (stmt == null || stmt.getOrigStmt() == null || stmt.getOrigStmt().originStmt == null) {
+            return null;
+        }
+
+        OptExpression rewritten = rewriteByTextMatch(input, context, parseNode);
```

Comment:
do you consider to try to reuse the MaterializedViewRewriter?
it has done the column mapping, new plan construction, add projection and so on.
maybe can be easier to support column reorder, expression and so on.

Reply:
It will be  a great work. But I think this will be rule based rewrite framework rather than text based rewrite framework. Text based rewrite is simple and limited in some scopes, but it can be easy for use.

I will think about this after this pr.